jerry-core/ecma/operations/ecma-reference.c

```diff
@@ -132,38 +183,66 @@ ecma_op_resolve_reference_value (ecma_object_t *lex_env_p, /**< starting lexical
       ecma_object_t *binding_obj_p = ecma_get_lex_env_binding_object (lex_env_p);
 
 #if ENABLED (JERRY_LCACHE)
-      ecma_property_t *property_p = ecma_lcache_lookup (binding_obj_p, name_p);
-
-      if (property_p != NULL)
+#if ENABLED (JERRY_ES2015)
+      /* Global environment record does not have withEnvironment flag,
+         therefore the direct lcache lookup can be performed.  */
+      if (lex_env_p == ecma_get_global_environment ())
       {
-        ecma_property_value_t *prop_value_p = ECMA_PROPERTY_VALUE_PTR (property_p);
+#endif /* ENABLED (JERRY_ES2015) */
+        ecma_property_t *property_p = ecma_lcache_lookup (binding_obj_p, name_p);
 
-        if (ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDDATA)
+        if (property_p != NULL)
         {
-          return ecma_fast_copy_value (prop_value_p->value);
-        }
+          ecma_property_value_t *prop_value_p = ECMA_PROPERTY_VALUE_PTR (property_p);
 
-        JERRY_ASSERT (ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
+          if (ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDDATA)
+          {
+            return ecma_fast_copy_value (prop_value_p->value);
+          }
 
-        ecma_getter_setter_pointers_t *get_set_pair_p = ecma_get_named_accessor_property (prop_value_p);
+          JERRY_ASSERT (ECMA_PROPERTY_GET_TYPE (*property_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
 
-        if (get_set_pair_p->getter_cp == JMEM_CP_NULL)
-        {
-          return ECMA_VALUE_UNDEFINED;
-        }
+          ecma_getter_setter_pointers_t *get_set_pair_p = ecma_get_named_accessor_property (prop_value_p);
+
+          if (get_set_pair_p->getter_cp == JMEM_CP_NULL)
+          {
+            return ECMA_VALUE_UNDEFINED;
+          }
 
-        ecma_object_t *getter_p = ECMA_GET_NON_NULL_POINTER (ecma_object_t, get_set_pair_p->getter_cp);
+          ecma_object_t *getter_p = ECMA_GET_NON_NULL_POINTER (ecma_object_t, get_set_pair_p->getter_cp);
 
-        ecma_value_t base_value = ecma_make_object_value (binding_obj_p);
-        return ecma_op_function_call (getter_p, base_value, NULL, 0);
+          ecma_value_t base_value = ecma_make_object_value (binding_obj_p);
+          return ecma_op_function_call (getter_p, base_value, NULL, 0);
+        }
+#if ENABLED (JERRY_ES2015)
       }
+#endif /* ENABLED (JERRY_ES2015) */
 #endif /* ENABLED (JERRY_LCACHE) */
 
       ecma_value_t prop_value = ecma_op_object_find (binding_obj_p, name_p);
 
       if (ecma_is_value_found (prop_value))
       {
-        return prop_value;
+#if ENABLED (JERRY_ES2015)
```

Comment:
This is a "huge" code block to guard and only a single line is the same in both cases. I would suggest using duplicating that single line and have a #else block of the #if.